[DDW-544] Use PopOver For All Stake Pool Tooltips

DominikGuzei opened the pull request:
This PR refactors all stake pool tooltips to use the PopOver component.

There are two pre-requisite PRs:
https://github.com/input-output-hk/daedalus/pull/2371
https://github.com/input-output-hk/react-polymorph/pull/160

---

## Todos
- [x] Fix issues with `NumericInput` which have been introduced by https://github.com/input-output-hk/react-polymorph/pull/159
- [x] Reduce stake pool hover timeout on delegation wizard to make it snappier
- [x] Fix "copy id" hover box width
- [x] Fix sometimes displaced tooltips
- [x] Fix popover covering "back to top" button
- [x] Look into rendering performance issues of stake pools grid
- [x] Try virtualized grid component to speedup grid rendering

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1613592296266900)
- [ ] All stake pool tooltips and the UX should be the same as before
- [ ] NumericInputs should behave the same way as before (the upgrade to latest react-polymorph also includes some bug fixes / changes to NumericInput)

---
###  Test Cases

DDW-544 Use PopOver component for all ThumbPool tooltips

```Gherkin
Scenario 1: Validate Stake Pool Popover

Given that I am on <Stake Pool View>
When I <action> a stake pool on the list
Then a popover is visible over the tile
And it contains all the data from the stakepool
And it's displayed at the front of all elements

|Stake Pool View|action|
|Delegation center|hover|
|Step 2 of the Delegation Wizard|hover|
|Stake pool tile view|click|
|Stake pool list view|click|
```

## Review Checklist

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

nikolaglumac commented:
@ManusMcCole @miorsufianiohk @yakovkaravelov please review/test this one today 🙏

nikolaglumac commented:
@alexander-rukin please test this one 🙏

nikolaglumac reviewed (changes requested):
@DominikGuzei I have posted a video recording of the issues I've found on the Slack thread.

There are multiple issues:

- sharp corners on the tooltip (should be rounded)
- small explanation tooltips show at weird places (top left of the screen) sometimes
- stake pools screen is now very very slow
- delegation wizard is very very slow

Thanks for fixing these 🙏

nikolaglumac commented:
@alexander-rukin @miorsufianiohk please focus on this one. I would like you to compare the performance of this PR versus the latest production Daedalus Mainnet 3.3.0. Thanks!

nikolaglumac commented:
@yakovkaravelov please review this one today 🙏

nikolaglumac reviewed (changes requested):
@DominikGuzei nice work! 

I don't face any more issues with performance (@alexander-rukin and @miorsufianiohk will also test and share their results), but I do still see the issue with small tooltips positions 😞 
Also, the "Stake pool ID" doesn't expand to cover the full width of the tooltip.

Can you please investigate this? 🙏 

![1](https://user-images.githubusercontent.com/376611/108675571-46dba800-74e7-11eb-947a-66154afdfba3.png)
![2](https://user-images.githubusercontent.com/376611/108675580-48a56b80-74e7-11eb-8e05-6dda90e79645.png)
![3](https://user-images.githubusercontent.com/376611/108675583-493e0200-74e7-11eb-8f1a-50f84111dc59.png)

alexander-rukin commented:
For me wizzard is still a bit slow. 
---
Checked the current prod version... 😱 performance on the wizard is broken 🤯 
So anything is already a good boost.

miorsufianiohk commented:
@DominikGuzei 
I tested and compared build 16675 against production 16556 and observed the tooltip is displayed on top of "Back to top" button. Happen on Step 2 of the Delegation Wizard,  Stake pool tile view, Stake pool list view. Please see: [Screenshot 1](https://drive.google.com/file/d/1JD2lQvkanyb11hER68Urb8xpqVEKeN7t/view?usp=sharing) , [Screenshot 2](https://drive.google.com/file/d/1i9P-o-1Iwzzz0tNWjsMBnwWTYjtHooC5/view?usp=sharing) and [Screenshot 3](https://drive.google.com/file/d/1h4hkQRBiyFIO0ZZgjBLjkz-3A50Hnc_9/view?usp=sharing)

yakovkaravelov reviewed (approved):
@DominikGuzei Good work. 🏅 
Btw, I noticed hooks are used here.
@nikolaglumac Are we going to use hooks in this repo also?
What about using them everywhere?

nikolaglumac commented:
Yes - we should start using hooks in all the places.

miorsufianiohk reviewed (changes requested):
Hi @DominikGuzei . This is broken in delegation center. Tested on 16770. 
See [attachment](https://drive.google.com/file/d/12MpiaqSjH-xyORDJWZV5UBjfjWvXruVg/view?usp=sharing)

DominikGuzei commented:
@miorsufianiohk @nikolaglumac all issues should be fixed now 🎉 ready for final review

miorsufianiohk reviewed (approved):
Looks good to me. Great work @DominikGuzei . Tested on 16790. 👍

alexander-rukin reviewed (approved):
Looks fine

nikolaglumac commented:
What about the performance? How does this measure up to current Mainnet production performance? 
@miorsufianiohk @ManusMcCole @alexander-rukin

nikolaglumac reviewed (approved):
Great work @DominikGuzei!
It works blazing fast now 🚀 

Before we merge it, I would like to get a review from @tomislavhoracek!

alexander-rukin commented:
> What about the performance? How does this measure up to current Mainnet production performance?
> @miorsufianiohk @ManusMcCole @alexander-rukin

The latest production is VERY slow, now with no delays, it looks much better and brings pleasure when using :D

tomislavhoracek commented on source/renderer/app/components/staking/delegation-center/WalletRow.js:
```diff
@@ -364,38 +346,31 @@ export default class WalletRow extends Component<Props, WalletRowState> {
+                        color={stakePoolRankingColor}
+                        currentTheme={currentTheme}
+                        onClose={() =>
+                          this.setState({
+                            highlightedPoolId: false,
+                          })
+                        }
+                        onOpen={() =>
+                          this.setState({
+                            highlightedPoolId: true,
+                          })
```
@DominikGuzei Can you isolate this to functions like `handleOpen` and `handleClose`. Not necessary but maybe more descriptive

tomislavhoracek commented on source/renderer/app/components/staking/stake-pools/StakePoolsList.js:
```diff
@@ -1,164 +1,91 @@
+ * Utility function to programmatically hide all pop overs
+ * This is used to hide the pool tooltips on scrolling the list
+ */
+function hideAllPopOvers() {
```
@domenkozar  Can you change this to an arrow function or we should use this as a new approach?

nikolaglumac replied:
You wanted to tag @DominikGuzei here I believe :-)

tomislavhoracek replied:
hehe that is true 😄

DominikGuzei replied:
@tomislavhoracek in this case there is no reason to use an arrow function as this is a global function and no context binding is required 😉 I don't think we have a policy of "only use arrow functions" and I would rather recommend the opposite: use arrow functions if it makes sense (e.g: inline definitions, context binding etc.)

nikolaglumac replied:
Agreed @DominikGuzei 👍

tomislavhoracek commented on source/renderer/app/components/staking/widgets/ThumbPool.js:
```diff
@@ -1,162 +1,85 @@
+    styles.content,
+    isDisabled ? styles.disabled : null,
+    isHighlighted ? styles.isHighlighted : null,
+    props.highlightOnHover ? styles.isOnHover : null,
```
Maybe `isHovered` or just `hovered` would be better than `isOnHover`

DominikGuzei replied:
I will rename this to `shouldHighlightOnHover` – this is not "the hover state" but rather if this style is added it sets up a `:hover` selector which in turn handles the "is hovered" automatically via css

tomislavhoracek replied:
👍

tomislavhoracek reviewed (commented):
@DominikGuzei AMAZING stuff, I really like new things with hooks and flow type Functions declaration you introduced here 👏  

I left just a few comments, really minor things that can be changed but not necessary.

darko-mijic commented:
Are there any noticeable changes for the end-users with this PR? Or the tooltips behave exactly as before?

gabriela-ponce commented:
@DominikGuzei Overall, this works fine, but I have a few concerns about this PR. Could these be improved?
For context, I compared this PR to the previous implementation of the stake pools popover and I found a few differences:

1 - Before, the popover was displayed at the bottom of the tile or row if there was not enough space to display it next to it (except on the delegation wizard, where this was not implemented). Now it's always displayed at the side, and sometimes it looks off.
**Before**: [Screenshot 1](https://drive.google.com/file/d/1vB135G3NO-jGZCdvlHr8g_Dkhj0Gfxhn/view?usp=sharing), [screenshot 2.](https://drive.google.com/file/d/1InaTG2P_L5reze2r2Azu_5MMJGrMHoyQ/view?usp=sharing)
**After**: [Screenshot 1](https://drive.google.com/file/d/1Rg6rT3lF27SuY0Mhr0sDXK0TC4jC5afW/view?usp=sharing), [screenshot 2](https://drive.google.com/file/d/1C9nJYjhuL5c_6N9YmQCfitNKibbC-IU3/view?usp=sharing), [screenshot 3](https://drive.google.com/file/d/1d3MQ8_zKi_SjNWXxWfxUpERsQ7obVwe-/view?usp=sharing)

2 - "Back to the top" was shown on top of the popovers, as implemented on PR #1468. You can check the previous implementation [here](https://drive.google.com/file/d/1lMqXp6EmkBHUZ16MGwe0AiIlxvK2ZSFG/view?usp=sharing).

3 - Also, the behavior of the popover when scrolling [was different](https://drive.google.com/file/d/1NzR8Fg7pkieN2LXs73h0SxchiA4ky-Sa/view?usp=sharing) as the popover moves with the tile and doesn't close until the user clicks away. [This is how it looks now.](https://drive.google.com/file/d/1VBxPv0nDVe-8CWOKNxlVpwBoRQ2YRvHC/view?usp=sharing)

4 - The performance is ok, but it's also slower [than before](https://drive.google.com/file/d/1voMx0lAT3UqQHuNA3W7jvVqzxOpmlxdH/view?usp=sharing). Check [this video](https://drive.google.com/file/d/1mgcboQmqvAWgx79PZJ9AmspTTeGZ39j8/view?usp=sharing) for the current implementation.

DominikGuzei commented:
@gabriela-ponce thanks for your review! I pushed several fixes and some of your observations should be checked by @darko-mijic or @nikolaglumac as they are "features" or already discussed:

1. I changed the placement to "auto" mode (we are using a 3rd party library for automatic popover placement, so we don't have 100% control anymore) … this resolves the issue you pointed out by always showing the popover at the side where there is the most space for it.
2. This was already discussed with @nikolaglumac and we decided that it's better to show the tooltip on top of the "back to top" because the tooltip is something the user explicitly wanted to see via interaction and covering that tooltip with an action the user might not want to perform (back to top) could even be considered a bug.
3. This would be easy to remove as I added it as a feature (hide tooltip on scroll) but I would like @darko-mijic and @nikolaglumac to decide what they prefer
4. The "slow" performance was just caused by the fade in/out animation - I removed the fade-in and left the fade-out at 100ms, feels more performant now 👍

nikolaglumac commented:
> 1. I changed the placement to "auto" mode (we are using a 3rd party library for automatic popover placement, so we don't have 100% control anymore) … this resolves the issue you pointed out by always showing the popover at the side where there is the most space for it.

This looks good 👍 

> 2. This was already discussed with @nikolaglumac and we decided that it's better to show the tooltip on top of the "back to top" because the tooltip is something the user explicitly wanted to see via interaction and covering that tooltip with an action the user might not want to perform (back to top) could even be considered a bug.

Agreed. The tooltip should show **over** the "Back to top" button.

> 3. This would be easy to remove as I added it as a feature (hide tooltip on scroll) but I would like @darko-mijic and @nikolaglumac to decide what they prefer

The tooltip should hide on the scroll.

> 4. The "slow" performance was just caused by the fade in/out animation - I removed the fade-in and left the fade-out at 100ms, feels more performant now 👍

I would like to hear some proper measurements and comparisons between the Daedalus Mainnet 3.3.1 and this PR's build.

nikolaglumac commented:
I would say this PR's build is faster performance-wise than the develop/production builds.
@gabriela-ponce @miorsufianiohk @ManusMcCole I would like to get some numbers from you so that we can compare the performance.

DominikGuzei commented:
@nikolaglumac regarding 4. there should be no difference in performance now for showing the tooltip. I fixed the performance issue a while ago by only initiating the Tippy instances when hovering the pool tiles - so there is always only one Popover regardless of the number of tiles

nikolaglumac commented:
Perfect @DominikGuzei 👌 
I would still like to hear @gabriela-ponce's opinion 🙏

gabriela-ponce commented:
@DominikGuzei @nikolaglumac Here's my feedback for the latest fixes:

> 1. I changed the placement to "auto" mode (we are using a 3rd party library for automatic popover placement, so we don't have 100% control anymore) … this resolves the issue you pointed out by always showing the popover at the side where there is the most space for it.

This fix looks good!

> 2. This was already discussed with @nikolaglumac and we decided that it's better to show the tooltip on top of the "back to top" because the tooltip is something the user explicitly wanted to see via interaction and covering that tooltip with an action the user might not want to perform (back to top) could even be considered a bug.

Got it 👍 

> 3. This would be easy to remove as I added it as a feature (hide tooltip on scroll) but I would like @darko-mijic and @nikolaglumac to decide what they prefer

I'd say this is fine, but there is still a [small delay](https://drive.google.com/file/d/1Vft6wVqf5zLcr3ol4ES66J7h4a6nEXu1/view?usp=sharing) after scrolling before the tooltip is hidden, causing the tooltip to be displaced for a second. Could this be improved?

> 4. The "slow" performance was just caused by the fade in/out animation - I removed the fade-in and left the fade-out at 100ms, feels more performant now 👍 

This looks much better, here are a few videos on how it looked before ([tile](https://drive.google.com/file/d/18kPIA7bYZa867s__87m3gdFc8vD7HAo7/view?usp=sharing) view, [grid](https://drive.google.com/file/d/19YEmUD2hP7xXaFqgTyz_oYL6WOyiwnh0/view?usp=sharing) view) and how it looks now ([tile](https://drive.google.com/file/d/1X943ZK80lBYoxDfhxXgstlp9Lpau8iQv/view?usp=sharing) view, [grid](https://drive.google.com/file/d/1b3mgADAhO31OG26gbBX842Jw-QiSkN9L/view?usp=sharing) view).
The delay in this implementation is still longer than before, but I'd say this is acceptable.

However, there is a delay sometimes to show the tooltip after scrolling. Not sure if this is related to this issue, but sometimes the tooltip doesn't show up or takes a little longer. Check video [1](https://drive.google.com/file/d/1X943ZK80lBYoxDfhxXgstlp9Lpau8iQv/view?usp=sharing) and [2](https://drive.google.com/file/d/11ETbl6w4URcCpIe5B49M2zov2UQ8VHn9/view?usp=sharing)

nikolaglumac commented:
@DominikGuzei can you please see if this can be improved? 🙏 
Thanks!

DominikGuzei commented:
@gabriela-ponce can you re-test the scrolling delay issue? I could not reproduce the slower opening times after scrolling/opening and closing pools several times but I changed the fade-out to hide instantly, so this definitely speeds up the performance for the "close on scroll" feature

nikolaglumac commented:
@gabriela-ponce we need your approval to merge this one 🙏

gabriela-ponce commented:
> @gabriela-ponce can you re-test the scrolling delay issue? I could not reproduce the slower opening times after scrolling/opening and closing pools several times but I changed the fade-out to hide instantly, so this definitely speeds up the performance for the "close on scroll" feature

@DominikGuzei I think this looks better now on the tile and grid view. Was this also implemented on the stake pool view on the Delegation Wizard? The delay is more noticeable there. Check [this](https://drive.google.com/file/d/1D87TlEPrMfcYySkGMgqGY8rOQ9gQJUw6/view?usp=sharing) video.
Also, usually, it takes longer to display the tooltips on that screen.

nikolaglumac reviewed (changes requested):
The new implementation is unfortunately slower than the original one...
I will take a closer look tomorrow. We need to improve it.

DominikGuzei commented:
@nikolaglumac which part/aspect is slower or needs to be improved? Unfortunately, there is no "general" trick up in my sleeve to improve the allover performance but maybe we can find some improvement for specific cases. (e.g: I'm actually not quite sure why the delegation wizard is a bit slower 🤔)

nikolaglumac commented:
@DominikGuzei I have a few I would like to try...

DominikGuzei commented:
@nikolaglumac @gabriela-ponce ready for a final review 🎉 the stake pools lists should be much faster / smooth now since I have refactored them to virtualized lists (only render visible rows). This means that the performance will stay the same, no matter how many pools we might have in the future.

nikolaglumac reviewed (changes requested):
Great work @DominikGuzei 🎉 
The performance of the tile/grid view is much much better now!

There is one small issue with the background-shadow/glow being clip on top of the tile view:
![shadow](https://user-images.githubusercontent.com/376611/112448975-3b340900-8d53-11eb-94f5-9340c7b34c4b.png)

Opening a stake pool details and clicking on the "Delegate" button will throw an error and render blank screen:
![Screenshot 2021-03-25 at 10 05 48](https://user-images.githubusercontent.com/376611/112449236-76ced300-8d53-11eb-9605-d3c9a1098922.png)

Is it possible to implement virtualized list to the table view?
![Screenshot 2021-03-25 at 10 14 34](https://user-images.githubusercontent.com/376611/112449309-88b07600-8d53-11eb-8de1-ed0374faac89.png)

nikolaglumac commented:
@DominikGuzei disregard this one - it is caused by another PR and already broken in develop.
@tomislavhoracek is fixing it...

![112449236-76ced300-8d53-11eb-9605-d3c9a1098922](https://user-images.githubusercontent.com/376611/112611548-3f792880-8e1e-11eb-934f-3922c291e875.png)

tomislavhoracek commented:
> @DominikGuzei disregard this one - it is caused by another PR and already broken in develop.
> @tomislavhoracek is fixing it...
> 
> ![112449236-76ced300-8d53-11eb-9605-d3c9a1098922](https://user-images.githubusercontent.com/376611/112611548-3f792880-8e1e-11eb-934f-3922c291e875.png)

@DominikGuzei Fixed in https://github.com/input-output-hk/daedalus/pull/2461

nikolaglumac commented:
@DominikGuzei fix has been merged into develop. Please merge develop into your branch 🙏

DominikGuzei commented:
@nikolaglumac thanks for the update - I merged latest develop, will work on the clipped glow and look into virtualizing the table view too 👍

DominikGuzei commented:
@nikolaglumac I have looked into virtualizing the table view but this will be a lot of work and I suggest creating a separate ticket for it as this PR is getting to big and it would be a pitty to block the improvements already made.

nikolaglumac commented:
Thanks! Agreed - please create a new ticket for that 🙏

nikolaglumac reviewed (changes requested):
@DominikGuzei can we avoid this?
There is an empty space on the right screen 
![Screenshot 2021-03-29 at 10 07 17](https://user-images.githubusercontent.com/376611/112806003-9842fe00-9076-11eb-8e61-9e951128b634.png)

DominikGuzei commented:
@nikolaglumac the empty space issue should be fixed now 👍

nikolaglumac reviewed (approved):
A.M.A.Z.I.N.G work @DominikGuzei as always 💅 🎉
